volatility3/framework/plugins/linux/kmsg.py

```diff
@@ -0,0 +1,387 @@
+    the CONFIG_LOG_BUF_SHIFT value as a power of 2.
+    This can also be modified by the log_buf_len kernel boot parameter.
+    In SMP systems with more than 64 CPUs this ringbuffer size is dynamically
+    allocated according the number of CPUs based on the value of
+    CONFIG_LOG_CPU_MAX_BUF_SHIFT, and the log_buf pointer is updated
+    consequently to the new buffer.
+    In that case, the original static buffer in __log_buf is unused.
+    """
+    @classmethod
+    def symtab_checks(cls, vmlinux) -> bool:
+        return vmlinux.has_type('printk_log')
+
+    def get_text_from_printk_log(self, msg) -> str:
+        msg_offset = msg.vol.offset + self.vmlinux.get_type('printk_log').size
+        return self.get_string(msg_offset, msg.text_len)
+
+    def get_log_lines(self, msg) -> Generator[str, None, None]:
+        if msg.text_len > 0:
+            text = self.get_text_from_printk_log(msg)
+            yield from text.splitlines()
+
+    def get_dict_lines(self, msg) -> Generator[str, None, None]:
+        if msg.dict_len == 0:
+            return None
+        dict_offset = msg.vol.offset + self.vmlinux.get_type('printk_log').size + msg.text_len
+        dict_data = self._context.layers[self.layer_name].read(dict_offset, msg.dict_len)
+        for chunk in dict_data.split(b'\x00'):
+            yield " " + chunk.decode()
+
+    def run(self) -> Iterator[Tuple[str, str, str, str, str]]:
+        log_buf_ptr = self.vmlinux.object_from_symbol(symbol_name='log_buf')
+        if log_buf_ptr == 0:
+            # This is weird, let's fallback to check the static ringbuffer.
+            log_buf_ptr = self.vmlinux.object_from_symbol(symbol_name='__log_buf').vol.offset
+            if log_buf_ptr == 0:
+                raise ValueError("Log buffer is not available")
+
+        log_first_idx = int(self.vmlinux.object_from_symbol(symbol_name='log_first_idx'))
+        cur_idx = log_first_idx
+        end_idx = log_first_idx  # We don't need log_next_idx here. See below msg.len == 0
+        while True:
+            msg_offset = log_buf_ptr + cur_idx  # type: ignore
+            msg = self.vmlinux.object(object_type='printk_log', offset=msg_offset)
+            if msg.len == 0:
+                # As per kernel/printk/printk.c:
+                # A length == 0 for the next message indicates a wrap-around to
+                # the beginning of the buffer.
+                cur_idx = 0
+            else:
+                facility, level, timestamp, caller = self.get_prefix(msg)
+                level_txt = self.get_level_text(level)
+                facility_txt = self.get_facility_text(facility)
+
+                for line in self.get_log_lines(msg):
+                    yield facility_txt, level_txt, timestamp, caller, line
+                for line in self.get_dict_lines(msg):
+                    yield facility_txt, level_txt, timestamp, caller, line
+
+                cur_idx += msg.len
+
+            if cur_idx == end_idx:
```

Comment:
Again, feels like a C way of breaking a loop rather than a python way.  Could we move the guard statement to become part of the while statement please?

Reply:
done, see the other comment